keps/0013-pull-based-audit-events-stream.md

```diff
@@ -0,0 +1,199 @@
+      * [Story 1](#story-1)
+      * [Story 2](#story-2)
+    * [Risks and Mitigations](#risks-and-mitigations)
+* [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+* [Drawbacks [optional]](#drawbacks-optional)
+* [Alternatives [optional]](#alternatives-optional)
+
+[Tools for generating]: https://github.com/ekalinin/github-markdown-toc
+
+## Summary
+
+This proposal aims at providing a pull-based stream for audit logging. This would look similar to a watch request to the API server, and provide stream of audit events. The stream does not return any previous events (i.e. no associated storage), it would only stream events that occurred after the stream was opened. Two non resource-url endpoints are installed to provide the stream. '/audits/' for all audit events generated in the API server, and '/audits/{namespace}' for audits events in the specified namespace. And query parameters could be used to get subset of audit events.
+
+## Motivation
+The current audit backends (log & webhook) require very high privileges (file access to the master) or prior configuration (webhook). There are a number of use cases for dynamic and unprivileged access to the audit logs. Examples include:
+- allow tenants to get audit events in their own namespace
+- provide a subset of audit events, so it will be easier to query and analyze (e.g. https://github.com/kubernetes/kubernetes/issues/56683)
+
+
+### Goals
+- Provide a new pull based audit event stream, which returns events that occurred after the stream was opened.
+- Allow users go get a subset of audit events.
+- Access control. Allow unprivileged users to get audits in their own namespace.
+- Bandwidth limit in server scope or per-namespace.
+
+### Non-Goals
+- Access control at user/group level(e.g. allow user A to only get audits events from users A,B,C)
+
+
+## Proposal
+
+### Make audit backends dynamicly registerable
+
+```golang
+// Registry is a decorator for the audit backend, it allows caller to dynamicly register/unregister new backends to it.
+type Registry interface {
+       Backend
+
+       // Register register a new Backend to the registry.
+       Register(Backend)
+
+       // UnRegister remove the Backend from registry.
+       UnRegister(Backend)
+}
+```
+When a request comes, a new backend is registered to the registry. Audit Events sent to this backend will be sent to users in the end.
+
+### endpoints and query parameters
+
+|endpoint            |definition                                  |supported paramsters                                 |
+|--------------------|--------------------------------------------|-----------------------------------------------------|
+|/audits             |all audit events generated in the API server|username, group, namespace, apiGroup, resource, verbs|
+|/audits/{namespaces}|audits events in the specified namespace    |username, group, apiGroup, resource, verbs           |
```

Comment:
These should be pseudo-resources. Something like:
```
/apis/audit.k8s.io/v1beta1/events
/apis/audit.k8s.io/v1beta1/namespaces/{namespace}/events
```
(I'm not sure how it works having both a cluster-scoped & namespace scoped version, but this is the basic idea)

Reply:
Initially, these should only support `watch`. But I could imagine expanding the scope of verbs someday.

Reply:
I will investigate this way.
Thanks.

Reply:
At least discovery will not work with cluster and namespace scoped resources, either or. So we need at least two resources. 

In theory we could serve `/apis/audit.k8s.io/v1beta1/events`, even though events are namespaced. This would mean though that each event has a namespace set, even those for cluster scoped resources.

For the other search criteria like "user" we could use labels, i.e. a label selector.